engines/sci/graphics/controls32.cpp

```diff
@@ -353,4 +366,303 @@ void GfxControls32::flashCursor(TextEditor &editor) {
 		_nextCursorFlashTick = g_sci->getTickCount() + 30;
 	}
 }
+
+
+reg_t GfxControls32::registerScrollWindow(ScrollWindow *scrollWindow) {
+	_lastScrollWindowId += 1;
+	_scrollWindows[_lastScrollWindowId.getOffset()] = scrollWindow;
+	return _lastScrollWindowId;
+}
+
+ScrollWindow *GfxControls32::getScrollWindow(reg_t id) {
+	Common::HashMap<int, ScrollWindow *>::iterator i;
+	i = _scrollWindows.find(id.getOffset());
+	if (i != _scrollWindows.end())
+		return i->_value;
+	else
+		return nullptr;
+}
+
+void GfxControls32::deregisterScrollWindow(reg_t id) {
+	_scrollWindows.erase(id.getOffset());
+}
+
+
+
+ScrollWindow::ScrollWindow(SegManager *segMan, const Common::Rect &rect,
+                           const Common::Point &point, reg_t plane, uint8 fore,
+                           uint8 back, GuiResourceId font, TextAlign align,
+                           uint8 border)
+: _plane(plane), _foreColor(fore), _backColor(back),
+  _borderColor(border), _fontId(font), _alignment(align), _rect50(rect),
+  _point74(point) {
+
+	_gfxText32 = new GfxText32(segMan, g_sci->_gfxCache);
+
+	_firstVisibleChar = 0;
+	_lastVisibleChar = 0;
+	_topVisibleLine = 0;
+	_bottomVisibleLine = 0;
+
+	_gfxText32->setFont(_fontId);
+
+	_fontScaledWidth = _gfxText32->_scaledWidth;
+	_fontScaledHeight = _gfxText32->_scaledHeight;
+
+	_visible = false;
+
+	_rect60.left = (_rect50.left * _fontScaledWidth) / 320;
+	_rect60.right = ((_rect50.right - 1) * _fontScaledWidth) / 320 + 1;
+	_rect60.top = (_rect50.top * _fontScaledHeight) / 200;
+	_rect60.bottom = ((_rect50.bottom - 1) * _fontScaledHeight) / 200 + 1;
+
+	_rect10.left = 2;
+	_rect10.top = 2;
+	_rect10.right = _rect60.width() - 2;
+	_rect10.bottom = _rect60.height() - 2;
+
+	_pointSize = _gfxText32->_font->getHeight();
+
+	uint8 skipColor = 0;
+	while (skipColor == _foreColor || skipColor == _backColor)
+		skipColor++;
+
+	assert(_rect60.width() > 0 && _rect60.height() > 0);
+	_bitmap = _gfxText32->createFontBitmap(_rect60.width(), _rect60.height(),
+	                                       _rect10, "", _foreColor, _backColor,
+	                                       skipColor, _fontId, _alignment,
+	                                       _borderColor, false, false);
+
+	debugC(1, kDebugLevelGraphics, "New ScrollWindow: rect size: %d x %d, bitmap: %04x:%04x", _rect10.width(), _rect10.height(), PRINT_REG(_bitmap));
+
+	_lastLineId = make_reg(0, 9999);
+
+	_screenItem = nullptr;
+}
+
+ScrollWindow::~ScrollWindow() {
+	// _gfxText32._bitmap will get GCed once ScrollWindow is gone.
+
+	delete _gfxText32;
+	delete _screenItem;
+}
+
+Common::Rational ScrollWindow::where() const {
+	return Common::Rational(_topVisibleLine, MAX(_numLines, 1));
+}
+
+void ScrollWindow::show() {
+	if (_visible)
+		return;
+
+	if (_screenItem == nullptr) {
+		CelInfo32 c;
+		c.bitmap = _bitmap;
+
+		ScaleInfo s;
```

Comment:
Could just put ScaleInfo() in the ScreenItem call itself and avoid the extra line in the code?